dataset/include/scipp/dataset/dataset_util.h

```diff
@@ -8,15 +8,67 @@
 
 namespace scipp::dataset {
 
-/// Iteration helper for generic code supporting data array and dataset.
-///
-/// Example:
-///     for (const auto &[name, data] : iter(dataarray_or_dataset)) {}
-template <class T> static decltype(auto) iter(T &d) {
-  if constexpr (std::is_same_v<T, Dataset>)
-    return d;
-  else
-    return d.iterable_view();
+[[noreturn]] inline void throw_nesting_error(const DataArray &) {
+  throw std::invalid_argument("Cannot assign DataArray, the right hand side "
+                              "contains a reference to the left hand side. "
+                              "Reference cycles are not allowed.");
+}
+
+[[noreturn]] inline void throw_nesting_error(const Dataset &) {
+  throw std::invalid_argument("Cannot assign Dataset, the right hand side "
+                              "contains a reference to the left hand side. "
+                              "Reference cycles are not allowed.");
+}
+
+template <class L, class R>
+void check_nested_in_assign(const L &lhs, const R &rhs);
+
+template <class T>
+void check_nested_in_assign(const T &lhs, const Variable &rhs) {
+  if (rhs.dtype() == dtype<DataArray>) {
+    for (const auto &nested : rhs.values<DataArray>()) {
+      check_nested_in_assign(lhs, nested);
+    }
+  } else if (rhs.dtype() == dtype<Dataset>) {
+    for (const auto &nested : rhs.values<Dataset>()) {
+      check_nested_in_assign(lhs, nested);
+    }
+  }
+}
+
+template <class T, class Key, class Value>
+void check_nested_in_assign(const T &lhs, const Dict<Key, Value> &rhs) {
+  for (const auto &[_, var] : rhs) {
+    check_nested_in_assign(lhs, var);
+  }
+}
+
+template <class L, class R>
+void check_nested_in_assign(const L &lhs, const R &rhs) {
+  if constexpr (std::is_same_v<L, R>) {
+    if (&lhs == &rhs) {
```

Comment:
Is this the endpoint of the recursive `check_nested_in_assign` calls? If so, it is not 100% clear to me (at the end of a workday) why pointer comparison of (e.g., DataArray?) is sufficient here. Wouldn't we also need to compare buffer pointers (`VariableConceptHandle`)? Or is that what is happening?

Reply:
As far as I understand it, storing DataArrays or Datasets in buffers copies them. So we can only get a cycle through constructs like the one in the PR description which are much harder to detect.